src/librustc_save_analysis/span_utils.rs

```diff
@@ -182,20 +182,31 @@ impl<'a> SpanUtils<'a> {
         let mut toks = self.retokenise_span(span);
         let mut prev = toks.real_token();
         let mut result = None;
+        let mut angle_count = 0;
         let mut bracket_count = 0;
         loop {
             let next = toks.real_token();
 
-            if (next.tok == token::Lt || next.tok == token::Colon) && bracket_count == 0 &&
+            if (next.tok == token::Lt || next.tok == token::Colon) &&
+               angle_count == 0 &&
+               bracket_count == 0 &&
                prev.tok.is_ident() {
                 result = Some(prev.sp);
             }
 
+            if bracket_count == 0 {
```

Comment:
So, this is a bit obscure. It seems like some comments might help to clarify what's going on.